Library/Source/NativeEngine.cpp

```diff
@@ -1327,6 +1365,47 @@ namespace Babylon
             static_cast<uint16_t>(height * backbufferHeight));
     }
 
+    Napi::Value NativeEngine::GetFramebufferData(const Napi::CallbackInfo& info)
+    {
+        bgfx::FrameBufferHandle fbh = BGFX_INVALID_HANDLE;
+        bgfx::requestScreenShot(fbh, "GetImageData");
+
+        while (s_bgfxCallback.m_screenShotBitmap.empty())
```

Comment:
This looks unsafe. Does `requestScreenShot` eventually call back on the same thread as this code? What probably should be happening is that the BgfxCallback::screenShot callback should post a message to the Babylon dispatcher somehow. Maybe NativeEngine should register with BgfxCallback so that BgfxCallback knows what objects need the callback. Currently, there might be race conditions here?

Reply:
No, a race condition here is not possible. Everything happens in the same thread. The screenshot callback is ran at the end of a bgfx::frame call. It happens a couple frames after the request. That's why there is a loop to wait for the data.

Reply:
This interface doesn't seem like it's intended to do what you are doing. Is there a way to render to a texture first and then call `readTexture`? If `requestScreenShot` is the best (only?) way to get frame buffer data, then I think I would abuse the filename parameter and store a callback instead. Then, there would be no need for the copy and storing of the screenshot.

Thoughts?

Reply:
best I think is to use a JS promise and pass the buffer as a JS napi bytearray. It would also be better when we try to do rendering in another thread.

Reply:
Okay, but still have to deal with the callback issue.

Reply:
I'll do a bgfx PR to add a void* userParam to the requestScreenshot method

Reply:
https://github.com/bkaradzic/bgfx/pull/2025

Reply:
I'm assuming branimir wants you to update bgfx code to make `char* filename` to a `void* data`? Can we do the filename hack for now and then update this code once bgfx is updated?